torch/quantization/fx/_equalize.py

```diff
@@ -224,3 +227,262 @@ def node_supports_equalization(node: Node, modules) -> bool:
+    """ Update all of the observer's equalization scale. For each
+    InputEqualizationObserver, we will find the location of the next
+    WeightEqualizationObserver, create it, and calculate the equalization scale
+    based on the two observers.
+
+    We will then return a dictionary mapping node names to the newly created
+    WeightEqualizationObserver.
+    """
+    weight_eq_obs_dict = {}
+    for node in model.graph.nodes:
+        if node.op == 'call_module' and isinstance(modules[node.target], _InputEqualizationObserver):
+            input_eq_obs = modules[node.target]
+            assert(isinstance(input_eq_obs, _InputEqualizationObserver))
+            next_node, weight_eq_obs = get_weight_eq_obs(node, model, modules)
+
+            if next_node is None or weight_eq_obs is None:
+                continue
+
+            weight_eq_obs(modules[str(next_node.target)].weight)
+
+            # Calculate and set the equalization scale values
+            equalization_scale = calculate_equalization_scale(input_eq_obs, weight_eq_obs)
+            input_eq_obs.set_equalization_scale(equalization_scale)
+            weight_eq_obs.set_equalization_scale(equalization_scale)
+
+            weight_eq_obs_dict[next_node.name] = weight_eq_obs
+
+    return weight_eq_obs_dict
+
+def convert_eq_obs(
+    model: GraphModule,
+    modules: Dict[str, nn.Module],
+    weight_eq_obs_dict: Dict[str, _WeightEqualizationObserver],
+) -> None:
+    """ Removes the input equalization observers and replaces them with mul
+    operators whenever applicable. Updates the input quantization observers with
+    the scaled input min/max values. Scales the weights by the current and next
+    equalization scales.
+
+    Before:
+                                     weight values
+                                          |
+                                      WeightEqObs
+                                          |
+                                     WeightQuantObs
+                                          |
+        x -> InpQuantObs -> InpEqObs -> linear -> OutQuantObs
+
+    After:
+                                              scaled weight values
+                                                      |
+                                                 WeightEqObs
+                                                      |
+       equalization scale                       WeightQuantObs
+              |                                       |
+        x -> mul -> InpQuantObs (scaled min/max) -> linear -> OutQuantObs
+    """
+    for node in model.graph.nodes:
+        if node.op == 'call_module' and isinstance(modules[node.target], _InputEqualizationObserver):
+            inp_quant_obs_node = node.args[0]
+            prev_node = inp_quant_obs_node.args[0]
+
+            # TODO: Possible special handling for connected linear layers
+            # Update the following input quantization observer's min/max values
+            scale_input_observer(node, modules)
```

Comment:
Can we add unit tests to make sure the scaled values in the observers are what we expect?

Reply:
Yup, would it be ok if I added an additional PR instead containing the new tests?

Reply:
I have added tests here! https://github.com/pytorch/pytorch/pull/60378